core/packages/contracts/src/Vault.sol

```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.19;
+
+import "openzeppelin/access/AccessControl.sol";
+import "./IVault.sol";
+
+contract Vault is IVault, AccessControl {
+    event Deposited(bytes indexed sovereign, uint256 amount);
+    event Withdrawn(bytes indexed sovereign, address recipient, uint256 amount);
+
+    error InsufficientBalance();
+    error ZeroAmount();
+    error CannotSendFunds();
+
+    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+    bytes32 public constant WITHDRAW_ROLE = keccak256("WITHDRAW_ROLE");
+
+    // Mapping of sovereign to balance
+    mapping(bytes => uint256) public balances;
+
+    constructor() {
+        _grantRole(ADMIN_ROLE, msg.sender);
+        _setRoleAdmin(WITHDRAW_ROLE, ADMIN_ROLE);
```

Comment:
ADMIN_ROLE is set to deployer account, and they can change withdrawal role even after deployment? Should have a revokeRole there in the contract / deployment process?

Reply:
See my latest PR to support upgrades:

https://github.com/Snowfork/snowbridge/pull/801

The deployer's admin privileges are revoked.